working/1426-extension-types/feature-specification.md

```diff
@@ -0,0 +1,937 @@
+from the context whether we are talking about the extension itself or a
+particular instantiation of a generic extension. For non-generic
+extensions, the on-type is the same in either case.
+
+An extension type `E` of the form
+<code>Ext<S<sub>1</sub>, .. S<sub>k</sub>></code>
+is a subtype of `Object?`, and a proper supertype of the
+instantiated on-type of `E`. In the case where the instantiated on-type of
+`E` is not a top type, `E` is also a proper subtype of `Object?`.
+
+*That is, the underlying on-type can only be recovered by an explicit cast
+(except when the on-type is a top type). So an expression whose type is an
+extension type is in a sense "in prison", and we can only obtain a
+different type for it by forgetting everything (going to a top type), or by
+means of an explicit cast, typically a downcast to the on-type.*
+
+When `E` is an extension type, a type test `o is E` or `o is! E` and a type
+check `o as E` can be performed. Such checks performed on a local variable
+can promote the variable to the extension type using the normal rules for
+type promotion.
+
+*Compared to the existing extension methods feature, there is no change to
+the type of `this` in the body of an extension type _E_: It is the on-type
+of _E_. Similarly, extension methods of _E_ invoked in the body of _E_ are
+subject to the same treatment as previously, which means that extension
+methods of the enclosing extension type can be invoked implicitly, and
+extension methods are given higher priority than instance methods on `this`
+when `this` is implicit.*
+
+
+### Prevent implicit invocations: Keyword 'type'
+
+This section specifies the effect of including the keyword `type` in the
+declaration of an extension type.
+
+*A reminder, to set the scene: An implicit extension member invocation is
+an invocation of the form `e.m()` (or similar for getter/setter/operator
+invocations) which invokes a member of an extension `E`, but where the
+static type of `e` is a type `T` that matches the on-type of `E`. This is
+the typical usage for the extension methods mechanism that Dart has had
+since version 2.6, and it is available with no changes for an extension
+declaration that does not include the keyword `type`. The other kinds of
+invocations of extension members are explicit extension member invocations
+(assuming that `E` is an extension type, an example would be
+`E(e).m()`) and typed extension member invocation (like `e.m()` where the
+static type of `e` is an extension type).*
+
+Let `E` be an extension declaration such that the keyword `extension` is
+followed by `type`. We say that `E` is an _explicit_ extension type
+declaration, and that it introduces an _explicit_ extension type.
+
+An explicit extension type declaration is not applicable for an implicit
+extension method invocation.
+
+*In other words, methods of an explicit extension type cannot be called on
+the on-type, only on the extension type. Otherwise, it works the same as an
```

Comment:
Except in the body of the extension itself, right?  Where they are implicitly treated as normal extension methods.

Reply:
Right, done.